staging/src/k8s.io/apimachinery/pkg/runtime/converter.go

```diff
@@ -205,13 +203,9 @@ func fromUnstructured(sv, dv reflect.Value) error {
 	}
 
 	// Check if the object has a custom JSON marshaller/unmarshaller.
-	if reflect.PtrTo(dt).Implements(unmarshalerType) {
```

Comment:
checking the ToUnstructured interface first seems reasonable, but don't we still need to check for a custom json marshaler as a fallback?

Reply:
`entry.CanConvertFromUnstructured` and `entry.ToUnstructured` (https://github.com/kubernetes-sigs/structured-merge-diff/pull/152/files#diff-fb99cda6711a615c2a07e578309552d9R164) handles both `ToUnstructured` and `json.Marshal`. It's quite a bit more efficient for `json.Marshal` as well since it caches the `reflect.Implements` checks, which are expensive.

Reply:
sorry, getting oriented... I see that is now baked into the SMD library.

Reply:
Some background: We have two caches used by reflectors to reason about structured types, one in runtime and another in SMD. Both are similar, they have a atomic reference to a copy-on-write map, and both are used to avoid doing redundant reflection operations on the structured types. I'm working on consolidating to use the cache in SMD since it's more comprehensive.